sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteGroupedRecordsToFiles.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.sdk.io.gcp.bigquery;
+
+import com.google.api.services.bigquery.model.TableRow;
+import java.util.UUID;
+import org.apache.beam.sdk.io.gcp.bigquery.BigQueryIO.Write;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.values.KV;
+
+/**
+ * Receives elements grouped by their (sharded) destination, and writes them out to a file.
+ * Since all the elements in the {@link Iterable} are destined to the same table, they are all
+ * written to the same file. Ensures that only one {@link TableRowWriter} is active per bundle.
+ */
+class WriteGroupedRecordsToFiles<DestinationT>
+    extends DoFn<KV<ShardedKey<DestinationT>, Iterable<TableRow>>,
+    WriteBundlesToFiles.Result<DestinationT>> {
+  private final String tempFilePrefix;
+
+  WriteGroupedRecordsToFiles(String tempFilePrefix) {
+    this.tempFilePrefix = tempFilePrefix;
+  }
+
+  @ProcessElement
+  public void processElement(ProcessContext c) throws Exception {
+    TableRowWriter writer = createWriter();
```

Comment:
I think this function should take more care to close the writer - closing it must be in a finally block. I *think* right now, upon visual inspection, there's no place where it can be unclosed, but if somebody adds a few more lines of code that might throw, the writer might not get closed.